python/GafferSceneTest/MergeScenesTest.py

```diff
@@ -0,0 +1,364 @@
+
+		# Combination of 2 input locations, hash can't be passed through.
+		self.assertNotEqual( merge["out"].attributesHash( "/sphere" ), sphere["out"].attributesHash( "/sphere") )
+		self.assertNotEqual( merge["out"].attributesHash( "/sphere" ), bigSphere["out"].attributesHash( "/sphere") )
+		# Single input location, hash should be passed through, and cache
+		# entries shared.
+		self.assertEqual( merge["out"].attributesHash( "/cube" ), cube["out"].attributesHash( "/cube") )
+		self.assertTrue(
+			merge["out"].attributes( "/cube", _copy = False ).isSame(
+				cube["out"].attributes( "/cube", _copy = False )
+			)
+		)
+
+		# Sets
+		# ====
+
+		# Sets are always merged
+
+		self.assertEqual( merge["out"].setNames(), IECore.InternedStringVectorData( [ "sphere", "both", "cube", "bigSphere" ] ) )
+		self.assertEqual( merge["out"].set( "sphere" ).value, IECore.PathMatcher( [ "/sphere" ] ) )
+		self.assertEqual( merge["out"].set( "cube" ).value, IECore.PathMatcher( [ "/cube" ] ) )
+		self.assertEqual( merge["out"].set( "both" ).value, IECore.PathMatcher( [ "/sphere", "/cube" ] ) )
+		self.assertEqual( merge["out"].set( "bigSphere" ).value, IECore.PathMatcher( [ "/sphere" ] ) )
+
+	def testGlobals( self ) :
+
+		a = GafferScene.CustomOptions()
+		a["options"].addChild( Gaffer.NameValuePlug( "test1", 1 ) )
+		a["options"].addChild( Gaffer.NameValuePlug( "test2", 2 ) )
+
+		b = GafferScene.CustomOptions()
+		b["options"].addChild( Gaffer.NameValuePlug( "test2", 4 ) )
+		b["options"].addChild( Gaffer.NameValuePlug( "test3", 3 ) )
+
+		merge = GafferScene.MergeScenes()
+		merge["in"][0].setInput( a["out"] )
+		merge["in"][1].setInput( b["out"] )
+
+		# Merging off by default
+
+		self.assertEqual( merge["out"].globals(), a["out"].globals() )
+		self.assertEqual( merge["out"].globalsHash(), a["out"].globalsHash() )
+
+		# Test merging
+
+		merge["globals"].setValue( True )
+
+		mergedGlobals = a["out"].globals()
+		mergedGlobals.update( b["out"].globals() )
+
+		self.assertEqual( merge["out"].globals(), mergedGlobals )
+
+	def testSingleInputPassThrough( self ) :
+
+		sphere = GafferScene.Sphere()
+		sphere["transform"]["translate"].setValue( imath.V3f( 1, 2, 3 ) )
+		sphere["sets"].setValue( "A" )
+		cube = GafferScene.Cube()
+
+		group = GafferScene.Group()
+		group["in"][0].setInput( sphere["out"] )
+		group["in"][1].setInput( sphere["out"] )
+
+		merge = GafferScene.MergeScenes()
+		merge["in"][0].setInput( group["out"] )
+
+		self.assertScenesEqual( merge["out"], group["out"] )
+		self.assertSceneHashesEqual( merge["out"], group["out"] )
+
+	def testNoInputsPassThrough( self ) :
+
+		merge = GafferScene.MergeScenes()
+		self.assertScenesEqual( merge["in"][0], merge["out"] )
+		self.assertSceneHashesEqual(
+			 merge["in"][0], merge["out"],
+			 # Can't expect transform hash to be identical, because
+			 # `SceneNode::hashTransform()` isn't called for the root.
+			 checks = self.allSceneChecks - { "transform" }
+		)
+
```

Comment:
Be good to add a no-first-input test.

Reply:
6e45cfeb2933d3e5e564dd84a7c964b51d1c883c